Add multiple selection in diff view

cruessler opened the pull request:
This is a first attempt at implementing multiple selection in the diff view.
You can use `Shift+ArrowUp` as well as `Shift+ArrowDown` to extend the
selection.

Known issue: If the selection spans more than one hunk, only the hunk that
contains the selection’s start is selected. I don’t know a good way to address
that. Preventing the selection from spanning more than hunk seems like an
obvious choice, but then, of course, you could only copy lines out of one hunk
at a time.

Still left to do is the actual copying to clipboard, but that should be fairly
easy once the above issue is resolved.

Closes #229.

extrawurst commented:
looks already like a great start. only two minor things:

1) is it intended that when u do both `up` and `down` that it simply grows the selection in both directions instead of moving the bottom like usual?
2) similar to 1) I think the scrolling should still work and go with the movable part of the selection
3) I think after stopping to do selection `up`/`down` without shift it should land where the movable part of the selection was

cruessler commented:
I have updated the PR.

ad 1) I found the initial behavior unexpected, too. I changed it so that `Shift+Arrow` now only moves the selection’s end which is closer to what you would expect if you come from vim.
ad 2), 3) I have not yet addressed that, but I’ll keep it in mind.

Doing the actual copying seems simple. At least on my machine (Ubuntu 20.04, X11), it worked immediately. Coming from vim, I have bound copy to `y`, but that can of course easily be changed.

extrawurst commented:
@cruessler I checked out the changes. works better now, but I would like to be able to also scroll up beyond the start selection line (basically inverting start/end selection index).
also I can confirm the copy to cb works on Mac too👍

cruessler commented:
@extrawurst I just updated the PR with another round of changes!

extrawurst commented on src/components/diff.rs:
```diff
@@ -113,37 +167,80 @@ impl DiffComponent {
+                .collect();
+
+            let mut ctx: ClipboardContext = ClipboardProvider::new()
+                .expect("failed to get access to clipboard");
```
when could this fail? are there reasonable cases? if so we could add a new error type and actually propagate it up, after all we are returning a result here.

cruessler replied:
I just had a look at the [source], and it looks like it can fail on Linux when there is no X11 session. I’ll add a new error type.

[source]: https://docs.rs/x11-clipboard/0.5.1/src/x11_clipboard/lib.rs.html#54

cruessler replied:
I changed it so that now the error is shown to the user. I tested that by having `copy_string` return a dummy error, but this is not included in the committed code.

extrawurst commented on src/components/diff.rs:
```diff
@@ -553,6 +657,10 @@ impl Component for DiffComponent {
                         }
                         Ok(true)
                     }
+                    keys::COPY => {
```
can u add the appropriate keys to the command (so that they show up on the bottom of the screen)?

extrawurst commented on src/components/diff.rs:
```diff
@@ -113,37 +167,80 @@ impl DiffComponent {
         Ok(())
     }
 
+    fn modify_selection(
```
one last nit-pick: if u look how editors like vscode handle this there is one more special case: when u select and scroll down and then let go of `shift` it depends on whether ur next input is `up` or `down` where `up will jump back to the top of the selection and `down` will simply move the current bottom selection one more down. scrolling up from the start of the selection of course works the opposite way around.

cruessler replied:
Done.

extrawurst replied:
works great now!🥳

extrawurst commented on .github/workflows/ci.yml:
```diff
@@ -31,6 +31,11 @@ jobs:
+    - name: Install dependencies for clipboard access
+      if: matrix.os == 'ubuntu-latest'
+      run: |
+        sudo apt-get -qq install libxcb-shape0-dev libxcb-xfixes0-dev
```
this is new third party dependency, people have to have installed right? how does this play with the release-linux-binary distribution? or are these libs that are present on every linux machine anyway?

cruessler replied:
`libxcb-shape0-dev` and `libxcb-xfixes0-dev` are the development files of `libxcb-shape0` and `libxcb-xfixes0`, and they need to be installed at compile time. I would assume that `libxcb-shape0` and `libxcb-xfixes0` which are required at runtime are present on any system that has X or Wayland installed (see `apt-cache rdepends libxcb-shape0 libxcb-xfixes0`).

If gitui is supposed to run on systems without X, we could put the parts related to clipboard access behind a flag and only conditionally compile them. What do you think?

extrawurst replied:
I am no expert on linux, does this mean we have to provide different built binaries for people who want to use gitui on server they just ssh into? (sorry if this is a stupid question)
as long as we can assume x/wayland to be installed for most people I am fine with this. I don't want to further complicate the deployment pipeline

cruessler replied:
No worries! I just realized that the binaries you are releasing are linked statically which makes the problem disappear because those binaries include all the necessary libraries and will run on almost any Linux system, regardless of the libraries installed on that system.

cruessler replied:
People will need those libraries only if they want to compile gitui themselves.

extrawurst replied:
Ok great! Then we are ready to go🥳

extrawurst commented:
@cruessler aside the linux question, can you rebase once more onto master, so we see if the windows pipeline (that got fixed now) runs through?

extrawurst commented:
Thanks so much @cruessler - great work!

cruessler commented:
@extrawurst One addition: I just saw that there are packages for Fedora and Arch. It is possible that they might need to be adapted to support clipboard access, but I am not familiar with building packages, so I don’t know how they work.

extrawurst commented:
@ignatenkobrain maintains fedora and @wezm did AUR maybe you guys can chip in before we release this and ruin your day? :)

wezm commented:
Looks like there's a couple of new runtime dependencies (`xcb` and friends). That's no big deal. 👍

ignatenkobrain commented:
I think it is fine for Fedora, but right now we are not able to build gitui because there is some bug that makes rustc to segfault on ppc64le.
